Upstream quality changes from Apex.AI part 1

MichaelOrlov opened the pull request:
This PR consolidates fixes related to the code quality and stability after backporting Rosbag2 PRs to Apex.AI. **PART 1**

### List of changes:
1. Fix for memory leaks in tests
   Details: Need to use custom deleter when releasing rcl serialized buffer to a shared pointer.
1. Move wait_for_playback_to_start() to public section
   - Rationale: Need to use in another package
   -  Also added timeout parameter for the wait_for_playback_to_start(..) to make it non-blocking call optionally and return boolean value.
1. Gracefully handle exceptions from on-play callbacks in player
  Rationale: Avoid abnormal termination in case of unhandled exceptions happening in callbacks.
1. Address Axivion warnings in rosbag2_py
   Description: 
   1. MisraC++2023-8.1.1 [required]: _transport.cpp:286 This variable shall not be implicitly captured in a lambda expression.; Field: []
   2. MisraC++2023-8.2.3 [required]: _transport.cpp:50 Cast removes const qualification; Field: [const char*->char*]
1. Fix the clang16 warning that the timestamp comparator is not initialized
1. Make sure to start with a clean subscriptions list in Recorder::record()
    Rationale: It could be some residual subscriptions in the list in case if the previous call to Recorder::record() failed for some reason and Recorder::stop() wasn't called.
1. Call discovery_future_.get() if it's ready in the `stop_discovery()`
1. Use absolute path instead of relative path in the `InfoEndToEndTestFixture`
1. Don't keep trying to publish the next message if we fail the first time. The `Player::play_next()` should return false right away if, for some reason, failed to publish the current message.

morlov-apexai committed:
Make sure to start with clean subscriptions list in Recorder::record()

- It could be some residual subscriptions in list in case if previous
call to Recorder::record() failed for some reason and Recorder::stop()
wasn't called.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Fix clang16 warning that timestamp comparator is not initialized

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Address Axivion warnings in rosbag2_py

1. MisraC++2023-8.1.1 [required]: _transport.cpp:286 This variable shall
 not be implicitly captured in a lambda expression.; Field: []
2. MisraC++2023-8.2.3 [required]: _transport.cpp:50
 Cast removes const qualification; Field: [const char*->char*]

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Gracefully handle exceptions from on-play callbacks in player

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Move wait_for_playback_to_start() to public section

- Rationale: Need to use in another package
- Added timeout parameter for the wait_for_playback_to_start(..) to make
it non-blocking call optionally and return boolean value.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Fix memory leaks in tests

- Use custom deleter when releasing rcl serialized buffer to a shared
pointer.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
There are a few failures on CI that need to be addressed. Turning this PR back to draft.  
Sorry it was premature to ask for review.

MichaelOrlov commented:
Ok. Now the CI is green and this PR is ready for review.

christophebedard commented on rosbag2_transport/src/rosbag2_transport/player.cpp:
```diff
@@ -196,9 +196,11 @@ class PlayerImpl
-  /// queue
-  void wait_for_playback_to_start();
+  /// \brief Waits on the condition variable until first message will be taken from read queue
+  /// @param timeout Maximum time in the fraction of seconds to wait for player to start.
```
Why "fraction of seconds"? We shouldn't need to care about whether `std::chrono::duration` is seconds, nanoseconds, etc. Users could just provide `std::chrono::nanoseconds(10)`, right?

MichaelOrlov replied:
Well, because this function is intended to be used and uses from unit tests.  
No one is going to use it with `wait_for_playback_to_start(std::chrono::nanoseconds(10))` because it doesn't make sense. The real wait time on CI ranges from hundreds of milliseconds to a few seconds, and I don't want to overcomplicate it with templates.
Therefore, a fraction of a second was chosen as the most suitable dimension for timeout.

christophebedard replied:
Even if no one is going to use `std::chrono::nanoseconds()`, the input type is `std::chrono::duration`, not `std::chrono::nanseconds` or `std::chrono::seconds`, so the method documentation should not treat it as "seconds." It should just treat it as a "duration," which could be anything (s/ms/ns/etc.), even if the caller really only wants to use seconds.

If it _only_ makes sense to provide a timeout in seconds, then the input type should be `std::chrono::seconds`.

MichaelOrlov replied:
According to the [Google](https://www.google.com/search?q=std%3A%3Achrono%3A%3Aduration%3Cdouble%3E&sca_esv=c7459735fc04b658&ei=RqqaZ5TaEvKgur8P_cTe6Qk&ved=0ahUKEwiUlb2F_JuLAxVykO4BHX2iN50Q4dUDCBM&uact=5&oq=std%3A%3Achrono%3A%3Aduration%3Cdouble%3E&gs_lp=Egxnd3Mtd2l6LXNlcnAiHXN0ZDo6Y2hyb25vOjpkdXJhdGlvbjxkb3VibGU-SNsnUJwPWPsgcAJ4AJABAJgBLKAB8wKqAQE5uAEDyAEA-AEBmAICoAIGwgIKEAAYsAMY1gQYR5gDAIgGAZAGCJIHATKgB-kH&sclient=gws-wiz-serp), ChatGPT and cppreference 
**std::chrono::duration<double> represents a time duration in seconds with floating-point precision.**
You can specify different time units using a second template parameter (e.g., std::milli, std::micro).
```
    std::chrono::duration<double> sec(2.5); // Represents 2.5 seconds

    // Convert to milliseconds
    std::chrono::duration<double, std::milli> millisec = sec;
    std::cout << "Duration: " << millisec.count() << " milliseconds" << std::endl;
```
I wouldn't change anything in this regard. IMO code and description are correct.

christophebedard replied:
It's not about being able to create a `std::chrono::duration<double>` object from a `double` value in seconds. That's just because the `std::chrono::duration` constructor accepts a `double` value, and, by default, it's in seconds. It's about being able to pass any kind of time value. For example, you can still do this:

```cpp
wait_for_playback_to_start(std::chrono::hours(1.0));
```

In which case it doesn't make sense to mention "seconds" in the documentation. That's it.

christophebedard replied:
anyway, if you prefer it this way, then alright

MichaelOrlov replied:
@christophebedard No, it doesn't work this way.
The `wait_for_playback_to_start(std::chrono::hours(1.0));` will not compile.
I have tried it from unit tests. It can't convert to the underlying type.
The build fails with an error message like:
```
/rosbag2_transport/test_play.cpp:987:60: error: no matching function for call to ‘std::chrono::duration<long int, std::ratio<3600> >::duration(double)’
  987 |   player->wait_for_playback_to_start(std::chrono::hours(1.0));
      |                                                            ^
```

However, I found mistakes in wording in the doxygen section for the `wait_for_playback_to_start`. I simply copy-pasted the description from the other `wait_for_playback_to_finish(..)` method and forgot to adopt it to start. 
Fixed it in the latest commit.

clalancette replied:
> The `wait_for_playback_to_start(std::chrono::hours(1.0));` will not compile.

That's true, but I think it is only true because I think that the type is over-specified to be a `std::chrono::duration<double>`.  If you changed the API to be either:

```
bool wait_for_playback_to_start(std::chrono::milliseconds timeout = std::chrono::milliseconds(-1))
```

or

```
bool wait_for_playback_to_start(std::chrono::duration<std::chrono::milliseconds::rep, std::milli> timeout = std::chrono::milliseconds(-1))
```

then you'd be able to pass arbitrary times in.  In other words, you probably don't need fractions of a tick, since you probably don't want to sleep for shorter than a millisecond.  In that case, you don't need the duration specified with a double, and you can instead use the "standard" types.

christophebedard replied:
> No, it doesn't work this way.
The `wait_for_playback_to_start(std::chrono::hours(1.0));` will not compile.

This compiles:

```cpp
  function(std::chrono::duration<double>(1.0));
  function(std::chrono::duration<double>(1.5));
  function(std::chrono::seconds(2));
  function(std::chrono::minutes(2));
  function(std::chrono::hours(2));
```

See: https://godbolt.org/z/vnc96cEr1

christophebedard commented on rosbag2_transport/src/rosbag2_transport/player.cpp:
```diff
@@ -1357,10 +1357,11 @@ bool PlayerImpl::publish_message(rosbag2_storage::SerializedBagMessageSharedPtr
-    run_play_msg_pre_callbacks(message);
     bool message_published = false;
     try {
+      // Calling on play message pre-callbacks
+      run_play_msg_pre_callbacks(message);
```
This should be in its own `try`/`catch` and log a `Failed to call on play message pre-callback on ...` error, otherwise an exception in `run_play_msg_pre_callbacks()` will be reported as a publish failure.

MichaelOrlov replied:
The idea was in a specific logic:
if the preconditions in terms of the pre-callbacks are not met, i.e. if an exception is thrown in one of the pre-callbacks,  then the publish method shouldn't be called.  
As far as I recall, we rely on this logic in one of our downstream packages, which uses the player directly.

christophebedard replied:
Good point, but this logic could still be achieved using separate `try`/`catch` blocks (e.g., using a simple `bool`) so that each one has its own error message. Something like this:

```cpp
bool do_publish = false;
try {
  run_play_msg_pre_callbacks(message);
  do_publish = true;
} catch (e) {
  // Log pre-callback error
}

if (do_publish) {
    try {
      // publish()
    } catch (e) {
      // Log publish error
  }
}
```

It has the same behaviour (skips publishing if pre-callback throws) but logs a helpful error message that directly mentions that a pre-callback threw.

MichaelOrlov replied:
Okay, I will rewrite it this way and will add a unit test for the case when `play_next` returns false if `pre_callback` throws.

christophebedard commented on rosbag2_py/src/rosbag2_py/_transport.cpp:
You could drop the references to Axivion or MISRA in the commit message/body when contributing here :grin:

christophebedard replied:
I'm referring to c5e49176d85c04252db76d91d163c1c6c6fc3c40

MichaelOrlov replied:
Why not? There is an explanation with reasonig why the changes were made. Otherwise, someone can start arguing or revert it.

christophebedard replied:
I simply meant that the changes have merit on their own; no need to mention where they're coming from. But this is just a comment, feel free to ignore.

christophebedard commented on rosbag2_transport/src/rosbag2_transport/recorder.cpp:
```diff
@@ -267,6 +267,7 @@ void RecorderImpl::record()
     throw std::runtime_error("No serialization format specified!");
   }
 
+  subscriptions_.clear();
```
Is this worth adding a test for?

MichaelOrlov replied:
I don't think so. It is more like defensive programming. No real issue so far.

christophebedard reviewed (changes requested):
Minor changes.

christophebedard commented:
Also, you don't really need to mention/frame the title of this PR as a "sync with Apex.AI"

morlov-apexai committed:
Use absolute path instead of relative path in the InfoEndToEndTestFixture

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Call discovery_future_.get() if it's ready in the `stop_discovery()`

- Also add missing `include <unordered_set>`

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Don't keep trying to publish next message if we failed first time

- The Player::play_next() should return false right away if for some
reason failed to publish current message.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Wrap run_play_msg_pre_callbacks(message) in its own try-catch

- Also added `play_next_returns_false_if_pre_callback_throw_exception`
unit test.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Move wait_for_playback_to_start(..) declaration and fixed wording

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@christophebedard Could you please give a one more round of review? 
I've addressed review comments and also added a few more fixes
> 7. Call discovery_future_.get() if it's ready in the stop_discovery()
> 8. Use absolute path instead of relative path in the InfoEndToEndTestFixture
> 9. Don't keep trying to publish the next message if we fail the first time. The Player::play_next() should return false right away if, for some reason, failed to publish the current message.

christophebedard reviewed (approved):
If you really want to proceed as-is with the `wait_for_playback_to_start()` docs, this looks good to me with green CI.

MichaelOrlov commented:
Pulls: ros2/rosbag2#1903
Gist: https://gist.githubusercontent.com/MichaelOrlov/09606b1727cd650b8ad81063eca5060f/raw/2aab4e80b4711be3bfd625549652c96db827c46d/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_transport rosbag2_py rosbag2_storage_mcap rosbag2_test_common rosbag2_tests
TEST args:  --packages-above rosbag2_transport rosbag2_py rosbag2_storage_mcap rosbag2_test_common rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/15128
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22460)](http://ci.ros2.org/job/ci_linux/22460/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16711)](http://ci.ros2.org/job/ci_linux-aarch64/16711/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1986)](http://ci.ros2.org/job/ci_linux-rhel/1986/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23247)](http://ci.ros2.org/job/ci_windows/23247/)

MichaelOrlov commented:
@christophebedard Thank you for review.

MichaelOrlov commented:
https://github.com/Mergifyio backport jazzy
